src/Rings/mpoly-local.jl

```diff
@@ -0,0 +1,366 @@
+###############################################################################
+
+function Base.show(io::IO, W::MPolyRing_loc)
+  print("Localization of the ", W.base_ring, " at the maximal ", W.max_ideal)
+end
+
+function Base.show(io::IO, w::MPolyElem_loc)
+  show(io, w.frac)
+end
+
+Nemo.symbols(R::MPolyRing_loc) = symbols(R.base_ring)
+Nemo.nvars(R::MPolyRing_loc) = nvars(R.base_ring)
+Nemo.parent(f::MPolyElem_loc) = f.parent
+Nemo.numerator(f::MPolyElem_loc) = numerator(f.frac)
+Nemo.denominator(f::MPolyElem_loc) = denominator(f.frac)
+
+elem_type(::MPolyRing_loc{T}) where {T} = MPolyElem_loc{T}
+elem_type(::Type{MPolyRing_loc{T}}) where {T} = MPolyElem_loc{T}
+parent_type(::Type{MPolyElem_loc{T}}) where {T} = MPolyRing_loc{T}
+
+###############################################################################
+# Arithmetics                                                                 #
+###############################################################################
+
+(W::MPolyRing_loc)() = MPolyElem_loc(W.base_ring(), W.max_ideal)
+(W::MPolyRing_loc)(i::Int) = MPolyElem_loc(W.base_ring(i), W.max_ideal)
+(W::MPolyRing_loc)(i::RingElem) = MPolyElem_loc(W.base_ring(i), W.max_ideal)
+(W::MPolyRing_loc)(f::MPolyElem) = MPolyElem_loc(f, W.max_ideal)
+(W::MPolyRing_loc)(g::AbstractAlgebra.Generic.Frac) = MPolyElem_loc(g, W.max_ideal)
+(W::MPolyRing_loc)(g::MPolyElem_loc) = W(g.frac)
+Base.one(W::MPolyRing_loc) = MPolyElem_loc(one(W.base_ring), W.max_ideal)
+Base.zero(W::MPolyRing_loc) = MPolyElem_loc(zero(W.base_ring), W.max_ideal)
+
++(a::MPolyElem_loc, b::MPolyElem_loc)   = MPolyElem_loc(a.frac+b.frac, a.parent.max_ideal)
+-(a::MPolyElem_loc, b::MPolyElem_loc)   = MPolyElem_loc(a.frac-b.frac, a.parent.max_ideal)
+-(a::MPolyElem_loc)   = MPolyElem_loc(-a.frac, a.parent.max_ideal)
+*(a::MPolyElem_loc, b::MPolyElem_loc)   = MPolyElem_loc(a.frac*b.frac, a.parent.max_ideal)
+==(a::MPolyElem_loc, b::MPolyElem_loc)   = a.frac == b.frac
+^(a::MPolyElem_loc, i::Int)    = MPolyElem_loc(a.frac^i, a.parent.max_ideal)
+
+function Oscar.mul!(a::MPolyElem_loc, b::MPolyElem_loc, c::MPolyElem_loc)
+  return b*c
+end
+
+function Oscar.addeq!(a::MPolyElem_loc, b::MPolyElem_loc)
+  return a+b
+end
+
+###############################################################################
+# Constructors for ideals                                                     #
+###############################################################################
+
+function singular_ring_loc(R::MPolyRing_loc{T}; ord::Symbol = :negdegrevlex) where T
+  return Singular.PolynomialRing(Oscar.singular_ring(base_ring(R.base_ring)),
+              [string(x) for x = Nemo.symbols(R)],
+              ordering = ord,
+              cached = false)[1]
+end
+
+mutable struct BiPolyArray_loc{S}
+  O::Array{S, 1}
+  S::Singular.sideal
+  Ox
+  Sx
+
+  function BiPolyArray_loc(Ox::T, b::Singular.sideal) where {T <: MPolyRing_loc}
```

Comment:
Is MPolyRing_loc parameterised by the elem_type S of T? If so you could use triangular dispatch here.

Perhaps you could also consider typing Ox and Sx above. e.g. Ox::MpolyRing_loc{S}

Reply:
I didn't type Ox and Sx, since it was not done in mpoly-graded.jl. i change it for the local rings.
MPolyRing_loc should be parametrized, like MPolyRing, by the type of the coefficient ring, for example the rationals, if the ring is Q[x,y]_<x,y>

Reply:
Yes, that's exactly the problem with not typing things; one ends up with the lack of typing propagating through the library. The result will be that the Julia compiler can't effectively optimise the code, resulting in slowdowns.

If it is considered practical, perhaps you could consider opening a ticket to type mpoly-graded.jl

Reply:
I agree that typing is reasonable. I will request the change or do it myself.